L1Trigger/L1THGCal/src/be_algorithms/HGCalShowerShape.cc

```diff
@@ -207,268 +209,118 @@ float HGCalShowerShape::sigmaPhiPhiTot(const l1t::HGCalMulticluster& c3d) const
 
 float HGCalShowerShape::sigmaEtaEtaMax(const l1t::HGCalMulticluster& c3d) const {
 
-    const edm::PtrVector<l1t::HGCalCluster>& clustersPtrs = c3d.constituents();
-
-    std::vector<int> layer ; // Size : ncl2D
-    std::vector<int> subdetID ;
-    std::vector<int> nTC ;
-    std::vector<float> tc_energy ; 
-    std::vector<float> tc_eta ;
-
-    for(const auto& clu : clustersPtrs){
-
-        layer.emplace_back(clu->layer());
-        subdetID.emplace_back(clu->subdetId());
+  std::map<int, vector<float> > tc_layer_energy;
+  std::map<int, vector<float> > tc_layer_eta;
 
-        const edm::PtrVector<l1t::HGCalTriggerCell>& triggerCells = clu->constituents();
-        unsigned int ncells = triggerCells.size();
-        nTC.emplace_back(ncells);
+  const edm::PtrVector<l1t::HGCalCluster>& clustersPtrs = c3d.constituents();
 
-        for(const auto& tc : triggerCells){
+  for(const auto& clu : clustersPtrs){
+    
+    int layer = HGC_layer(clu->subdetId(),clu->layer());    
+    
+    const edm::PtrVector<l1t::HGCalTriggerCell>& triggerCells = clu->constituents();
 
-            tc_energy.emplace_back(tc->energy());
-            tc_eta.emplace_back(tc->eta());
+    for(const auto& tc : triggerCells){
 
-        }
+      tc_layer_energy[layer].emplace_back(tc->energy());
```

Comment:
`tc_layer_energy` and `tc_layer_eta` should first be initialized. 
Before the TC loop:
```cpp
auto layer_energy_itr = tc_layer_energy.emplace(layer, 0).first;
```
and in the loop:
```cpp
layer_energy_itr->second.emplace_back(tc->energy());
```